yarn-project/aztec-rpc/src/account_state/account_state.ts

```diff
@@ -57,78 +58,98 @@ export class AccountState {
+    for (let i = 0; i < unverifiedDatas.length; ++i) {
+      const l2Block = l2Blocks[i];
+      const dataChunks = unverifiedDatas[i].dataChunks;
+
+      // Try decrypting the unverified data.
       const txIndices: Set<number> = new Set();
       const txAuxDataDaos: TxAuxDataDao[] = [];
-      for (let i = 0; i < dataChunks.length; ++i) {
-        const txAuxData = TxAuxData.fromEncryptedBuffer(dataChunks[i], this.privKey, this.grumpkin);
+      for (let j = 0; j < dataChunks.length; ++j) {
+        const txAuxData = TxAuxData.fromEncryptedBuffer(dataChunks[j], this.privKey, this.grumpkin);
         if (txAuxData) {
-          const txIndex = Math.floor(i / KERNEL_NEW_COMMITMENTS_LENGTH);
+          // We have successfully decrypted the data.
+          const txIndex = Math.floor(j / KERNEL_NEW_COMMITMENTS_LENGTH);
           txIndices.add(txIndex);
           txAuxDataDaos.push({
             ...txAuxData,
             nullifier: Fr.random(), // TODO
-            index: dataStartIndex + i,
+            index: dataStartIndex + j,
           });
         }
       }
 
-      if (txIndices.size) {
-        decrypted.push({ blockNo, txIndices: [...txIndices], txAuxDataDaos });
-        this.log(`Decrypted ${txIndices.size} tx aux data in block ${blockNo}.`);
-      } else {
-        this.log(`No tx aux data found in block ${blockNo}`);
-      }
-
+      blocksAndTxAuxData.push({ block: l2Block, userPertainingTxIndices: [...txIndices], txAuxDataDaos });
       dataStartIndex += dataChunks.length;
     }
 
-    if (decrypted.length) {
-      const txAuxDataDaos = decrypted.map(({ txAuxDataDaos }) => txAuxDataDaos);
-      await this.db.addTxAuxDataBatch(txAuxDataDaos.flat());
-
-      const blocks = await this.node.getBlocks(from, take);
-      const targetBlocks = decrypted.map(({ blockNo }) => blocks.find(b => b.number === blockNo)!);
-      const txIndices = decrypted.map(({ txIndices }) => txIndices);
-      await this.processBlocks(targetBlocks, txIndices, txAuxDataDaos);
-    }
+    await this.processBlocksAndTxAuxData(blocksAndTxAuxData);
 
-    this.syncedToBlock = toBlockNo;
+    this.syncedToBlock = l2Blocks[l2Blocks.length - 1].number;
+    this.log(`Synched block ${this.syncedToBlock}`);
   }
 
-  private async processBlocks(blocks: L2Block[], txIndices: number[][], txAuxDataDaos: TxAuxDataDao[][]) {
+  private async processBlocksAndTxAuxData(
+    blocksAndTxAuxData: { block: L2Block; userPertainingTxIndices: number[]; txAuxDataDaos: TxAuxDataDao[] }[],
+  ) {
+    const txAuxDataDaosBatch: TxAuxDataDao[] = [];
     const txDaos: TxDao[] = [];
-    for (let i = 0; i < blocks.length; ++i) {
-      const block = blocks[i];
-      txIndices[i].map((txIndex, j) => {
-        const txHash = getTxHash(block, txIndex);
+    for (let i = 0; i < blocksAndTxAuxData.length; ++i) {
+      const { block, userPertainingTxIndices, txAuxDataDaos } = blocksAndTxAuxData[i];
+      const blockHash = keccak(block.encode());
```

Comment:
Could we wrap the block in a BlockContext object and store lazy initialised values for block hash and the tx hashes? Just thinking we only want to produce those hashes once.